### PR TITLE
Decoding handle connection instance from protos to Kotlin Recipes

### PR DESCRIPTION
**Pull request by alxrsngrtn**
A continuation of #5613. This deserialized protos into Kotlin.

### CODE REVIEW

File: java/arcs/core/data/proto/ParticleProtoDecoder.kt [hunk 1]
```diff
@@ -38,7 +38,13 @@ fun HandleConnectionProto.decode(
     val recipeHandle = requireNotNull(context.recipeHandles[handle]) {
         "Handle '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
     }
-    return HandleConnection(handleSpec, recipeHandle)
+    val connectionType = try { type.decode() } catch (e: IllegalArgumentException) {
+        throw IllegalArgumentException(
```
Comment (piotrswigon) [inline java/arcs/core/data/proto/ParticleProtoDecoder.kt:42]: Why do we need to do this? When would a handle connection type be missing?

Comment (alxrsngrtn) [reply java/arcs/core/data/proto/ParticleProtoDecoder.kt:42]: It shouldn't be missing within normal usage. An error may be thrown when a developer writes or updates a test or is creating new functionality. In this case, error messaging is friendlier. I'm trying to program defensively. Ideally, the above cases wouldn't have nulls, but they will in exceptional cases. Here, we provide helpful messages. This is the line of thinking behind this exception handling logic.

Comment (piotrswigon) [reply java/arcs/core/data/proto/ParticleProtoDecoder.kt:42]: If we add rethrows with nicer messages for every field in the proto that may be missing (and everything may be missing in protos, everything is optional!) our code will grow 3-5x in size (LOC) while not providing benefit as exceptions already have stack traces that will tell you exactly where the exception happened. I would advise against adding such bloat.

File: java/arcs/core/data/Recipe.kt [hunk 2]
```diff
@@ -70,7 +71,7 @@ fun Recipe.Particle.toPlanParticle() = Plan.Particle(
 /** Translates a [Recipe.Particle.HandleConnection] into a [Plan.HandleConnection] */
 fun Recipe.Particle.HandleConnection.toPlanHandleConnection() = Plan.HandleConnection(
     mode = spec.direction,
-    type = spec.type,
+    type = type,
```
Comment (piotrswigon) [inline java/arcs/core/data/Recipe.kt:74]: nice! So you are indeed doing this already :)

File: javatests/arcs/core/data/RecipeTest.kt [hunk 3]
```diff
@@ -110,17 +110,20 @@ class RecipeTest {
             name = "ParticleName",
             location = "com.Particle",
             connections = listOf(
-                HandleConnectionSpec("data", HandleMode.Read, contactCollectionType)
+                HandleConnectionSpec("data", HandleMode.Read, personCollectionType)
             ).associateBy { it.name }
         )
 
+        // Given that the ParticleSpec type is different from the HandleConnection instance type,
+        // prefer the HandleConnection instance type for the translation.
```
Comment (alxrsngrtn) [inline javatests/arcs/core/data/RecipeTest.kt:118]: @piotrswigon I modified this test for the update. I can also add a new test, but I wanted to be a little more lofi.

**Pull request merged**